kdbx/src/main/java/org/linguafranca/pwdb/kdbx/KdbxKeyFile.java

```diff
@@ -52,7 +54,22 @@ public static byte[] load(InputStream inputStream) {
                 return null;
             }
             if (version.equals("2.0")) {
-                return Hex.decodeHex(data.replaceAll("\\s",""));
+               
+                byte[] hexData = Hex.decodeHex(data.replaceAll("\\s",""));
+               
+                MessageDigest md = Encryption.getSha256MessageDigestInstance();
+                byte[] computedHash = md.digest(hexData);
+               
+                String hashToCheck = (String) xpath.evaluate("//KeyFile/Key/Data/@Hash", doc, XPathConstants.STRING);
+                byte[] verifiedHash = Hex.decodeHex(hashToCheck);
+                
+                for(int i = 0; i < verifiedHash.length; i++) {
```

Comment:
Can use `Arrays.equals` here.

Also I think an exception should be thrown if the hash doesn't match. `IllegalStateException` will do ...

Reply:
Yeah, you're right!
As soon as possible, I will send you this update!

Reply:
Sorry for the previous comment.
I made a mistake:

- I can't use Arrays.equals because the computedHash array has a length > of the verifiedHash (see the hash in the keyfile). Using Arrays.equals implies copying the data from the computedHash to a tmp array, then calling the Arrays.equals method.
- I was undecided if throw an exception or return null, because the null data is checked by the caller's method

Giuseppe

Reply:
I think `Arrays.equals(Arrays.copyOf(computedHash, 4), verifiedHash)` works well enough. 

It should be an exception because the hash is not correct.